### PR TITLE
Hnimrama/ix atom

### PR DESCRIPTION
**Pull request by hnimra-amd**
# PR #238 — InferenceX ATOM W1 (MI300X perf gates)

**PR:** https://github.com/ROCm/cvs/pull/238  
**Head:** `hnimrama/IX-atom` → **Base:** `dev/dtni`

**Jira:** [AIMVT-236](https://amd-hub.atlassian.net/browse/AIMVT-236) · [AIMVT-244](https://amd-hub.atlassian.net/browse/AIMVT-244)

---

## Motivation

CVS needs a first-class **InferenceX ATOM** automation path aligned with the **DTNI Validation Tracker (IX ATOM)** — not the legacy `inferencemax_single` uplift. This PR:

- Introduces the **`inferencex_atom_single`** suite with an ATOM-native driver (`atom.entrypoints.openai_server` + `atom.benchmarks.benchmark_serving`).
- Ships **W1 DeepSeek R1 FP8** variant configs for **MI300X** and **MI355X** (perf, smoke, MTP3) with calibrated / CI-seeded thresholds.
- Closes **M1 / Phase A** on MI300X: W1 perf with `enforce_thresholds: true` after lab confirmation.
- Documents the IX-atom roadmap (W1–W18, accuracy, metric tiers, parity frameworks) in `plans/inferencex-atom-cvs-automation-plan.md`.

**MI355X** variant configs ship with `enforce_thresholds: false` until hardware is available — they do not block merge or MI300X milestone work.

**Base branch:** `dev/dtni`.

---

### Suite and orchestration

- Rename **`inferencemax_single` → `inferencex_atom_single`**; legacy `inferencemax/` configs removed.
- New **`InferenceXAtomJob`** (`inferencex_atom_orch.py`):
  - `params.driver=atom` → ATOM server + `benchmark_serving`; JSON artifacts parsed via `to_client_metrics`.
  - `params.driver=vllm` retained for interim GPT-OSS uplift variants only.
- Canonical config layout: flat sibling pairs under `cvs/input/config_file/inference/inferencex_atom_single/`:
  - `{gpu}_inferencex-atom-single_{model}_{precision}[_{mode}]_config.json`
  - `{gpu}_inferencex-atom-single_{model}_{precision}[_{mode}]_threshold.json`
- `schema_version: 1`, typed loader, and `ix_recipes.json` recipe pins (`dsr1-fp8-mi300x-atom`, etc.).
- Cluster examples: `mi300x_atom_single.json`, `mi355x_atom_single.json`; container names pinned (`inferencex_atom_mi300x` / `inferencex_atom_mi355x`).
- Shared suite helpers: `inference_suite_lifecycle.py`, `inference_suite_results_table.py`.

### W1 variants shipped

| Variant stem | Arch | `enforce_thresholds` | Notes |
|--------------|------|----------------------|-------|
| `mi300x_inferencex-atom-single_deepseek-r1_fp8_perf` | MI300X | `true` | M1 gate — ISL=OSL=1024, CONC=128/256, 1000 prompts |
| `mi300x_inferencex-atom-single_deepseek-r1_fp8_smoke` | MI300X | `false` | 128-prompt pre-gate |
| `mi300x_inferencex-atom-single_deepseek-r1_fp8_mtp3` | MI300X | `false` | MTP3 recipe |
| `mi355x_inferencex-atom-single_deepseek-r1_fp8_perf` | MI355X | `false` | CI-seeded thresholds (plan Section 4.3) |
| `mi355x_inferencex-atom-single_deepseek-r1_fp8_mtp3` | MI355X | `false` | CI-seeded |

### Threshold / metrics plumbing

- **Tiered gates** via `test_cell_metrics` (`METRIC_TIERS`: throughput, ttft, tpot, health, record) — one pytest row per tier per sweep cell.
- **`to_client_metrics`**: derive `client.failed` when ATOM omits it (`num_prompts - completed`), then compute `client.success_rate`; add `client.output_tput_per_gpu`.
- **Tpot tier** gates `p99_tpot_ms` (ATOM `benchmark_serving` emits p99 tails; `p95_tpot_ms` may be absent even with `metric_percentiles: "95,99"`). Tier enforcement skips metrics missing from the artifact.
- MI300X W1 perf thresholds recalibrated from **2026-06-25** lab run (throughput mins = measured × 0.9; latency maxes = measured × 1.1).

### Platform / shared infra (supporting changes)

- `cvs/lib/utils/` — shared `config_loader`, `verdict`, sweep selector.
- `vllm_single` suite refactor to the same lifecycle / metric pattern.
- Runbook: `cvs/input/config_file/inference/inferencex_atom_single/README.md`.

### Out of scope (follow-up on `dev/dtni`)

- M2 gsm8k accuracy
- M3 P1 workloads (W2, W3, W13, W17)
- M4 `inferencex_atom_vllm_single` / `inferencex_atom_sglang_single` parity frameworks
- M5 multi-node (prioritized immediately after M4 parity)

---

### CI / unit (no GPU)

- [x] `pytest cvs/tests/inference/inferencex_atom/ -q`
- [x] `pytest cvs/lib/inference/unittests/test_inferencex_atom_parsing.py -q`
- [x] `pytest cvs/lib/inference/unittests/test_inferencex_atom_config_loader.py -q`
- [x] Config loader / sweep selector / orch parse tests pass

### Lab — MI300X smoke

**Launcher:** `CTR-SVDT-L005` (`10.7.54.167`) · **GPU node:** `10.245.135.75`

```bash
cd ~/cvs && make install && source .cvs_venv/bin/activate

SMOKE_DIR=~/input/config_file/inference/inferencex_atom_single/smoke
mkdir -p "$SMOKE_DIR"
cvs copy-config inference/inferencex_atom_single/mi300x_inferencex-atom-single_deepseek-r1_fp8_smoke_config.json \
  --output "$SMOKE_DIR/mi300x_inferencex-atom-single_deepseek-r1_fp8_smoke_config.json"
cvs copy-config inference/inferencex_atom_single/mi300x_inferencex-atom-single_deepseek-r1_fp8_smoke_threshold.json \
  --output "$SMOKE_DIR/mi300x_inferencex-atom-single_deepseek-r1_fp8_smoke_threshold.json"

TS=$(date +%Y%m%d_%H%M%S)
HTML="$HOME/cvs_results/${TS}_ix-atom-smoke_mi300x.html"
LOG="$HOME/cvs_results/${TS}_ix-atom-smoke_mi300x.log"

cvs run inferencex_atom_single \
  --cluster_file ~/input/cluster_file/mi300x_atom_single.json \
  --config_file "$SMOKE_DIR/mi300x_inferencex-atom-single_deepseek-r1_fp8_smoke_config.json" \
  --html="$HTML" --self-contained-html --log-file="$LOG" -vvv -s
```

- [x] **11 passed**, 0 failed (~13 min)

### Lab — MI300X W1 perf (M1 gate)

```bash
cd ~/cvs && make install && source .cvs_venv/bin/activate

PERF_DIR=~/input/config_file/inference/inferencex_atom_single/perf
mkdir -p "$PERF_DIR"
cvs copy-config inference/inferencex_atom_single/mi300x_inferencex-atom-single_deepseek-r1_fp8_perf_config.json \
  --output "$PERF_DIR/mi300x_inferencex-atom-single_deepseek-r1_fp8_perf_config.json"
cvs copy-config inference/inferencex_atom_single/mi300x_inferencex-atom-single_deepseek-r1_fp8_perf_threshold.json \
  --output "$PERF_DIR/mi300x_inferencex-atom-single_deepseek-r1_fp8_perf_threshold.json"

TS=$(date +%Y%m%d_%H%M%S)
HTML="$HOME/cvs_results/${TS}_ix-atom-w1-perf_mi300x.html"
LOG="$HOME/cvs_results/${TS}_ix-atom-w1-perf_mi300x.log"

cvs run inferencex_atom_single \
  --cluster_file ~/input/cluster_file/mi300x_atom_single.json \
  --config_file "$PERF_DIR/mi300x_inferencex-atom-single_deepseek-r1_fp8_perf_config.json" \
  --html="$HTML" --self-contained-html --log-file="$LOG" -vvv -s
```

- [x] Lifecycle stages pass (container, model fetch, server start, benchmark client, teardown).
- [x] Both sweep cells: `CONC=128`, `CONC=256` (ISL=OSL=1024, TP=8); server reused across cells.
- [x] All metric tiers pass under `enforce_thresholds: true`.
- [x] HTML report attached to PR.

### MI355X

- [x] Not required for merge — configs ship; flip `enforce_thresholds` when hardware is available.

---

### MI300X lab — `mi300x_inferencex-atom-single_deepseek-r1_fp8_perf`

- **Branch:** `hnimrama/IX-atom` @ `07c90a7`
- **Launcher:** `CTR-SVDT-L005` (`10.7.54.167`) · **GPU node:** `10.245.135.75`
- **Install:** `make install` from repo root (`.cvs_venv`)
- **Image:** `rocm/atom-dev:latest`
- **Model:** `deepseek-ai/DeepSeek-R1-0528` (FP8, TP=8)
- **Outcome:** **17 passed, 0 failed** (~22 min)

| Cell | `client.output_throughput` (measured) | Threshold (min) | Result |
|------|----------------------------------------|-----------------|--------|
| CONC=128 | 2867 tok/s | 2580 tok/s | PASS |
| CONC=256 | 4697 tok/s | 4227 tok/s | PASS |

| Cell | TTFT / TPOT gates | Result |
|------|-------------------|--------|
| CONC=128 | mean TTFT 811 ms (max 892); p99 TTFT 6511 ms (max 7162); mean TPOT 42.5 ms; p99 TPOT 46.7 ms (max 51.4) | PASS |
| CONC=256 | mean TTFT 728 ms; p99 TPOT 59.7 ms (max 65.6) | PASS |

### MI300X lab — smoke

- **Outcome:** **11 passed, 0 failed** (~13 min)

### Artifacts (attach to PR)

[inferencex_atom_single_2026-06-25T175704.zip](https://github.com/user-attachments/files/29476631/inferencex_atom_single_2026-06-25T175704.zip)

### Unit tests

- [x] CI / local unit suite green on PR branch.

---

## Submission Checklist

- [x] PR open: [#238](https://github.com/ROCm/cvs/pull/238)
- [x] Jira linked: [AIMVT-236](https://amd-hub.atlassian.net/browse/AIMVT-236), [AIMVT-244](https://amd-hub.atlassian.net/browse/AIMVT-244)
- [x] Base branch is **`dev/dtni`**
- [x] Lab run used **`make install`** from this branch (not a stale site-packages install)
- [x] MI300X W1 perf + smoke HTML reports attached
- [x] `enforce_thresholds: true` only on MI300X W1 perf — confirmed in lab
- [x] MI355X variants left at `enforce_thresholds: false` (pending hardware)
- [x] Plan doc reviewed for milestone alignment
- [x] Reviewer aware M2 (gsm8k), M3 (W2/W3/W13/W17), and M5 (multi-node) are follow-ups on `dev/dtni`

### CODE REVIEW

File: cvs/lib/inference/utils/inferencex_atom_config_loader.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+
+class InferenceXAtomRoles(_Forbid):
+    server: InferenceXAtomRoleServer = InferenceXAtomRoleServer()
+
+
+class InferenceXAtomParams(_Forbid):
+    # ``atom`` uses ATOM openai_server + benchmark_serving; ``vllm`` keeps the
+    # interim uplift path (vllm serve + vllm bench serve).
+    driver: Literal["atom", "vllm"] = "vllm"
```
Comment (atnair-amd) [inline cvs/lib/inference/utils/inferencex_atom_config_loader.py:42]: **[MAJOR] `driver="vllm"` default has no guard — silent misrouting for new ATOM configs**

The two `gpt-oss-120b` configs intentionally omit `driver` and rely on this default (their serve args are vllm-specific — that is correct). But there is no safeguard for a new config author who adds an ATOM config and forgets the field: they silently get a different binary, different log paths, and a different client — all with exit 0. Suggest adding a Pydantic validator or a conftest-level log warning that fires when `framework == "inferencex_atom_single"` and `driver == "vllm"`, so the mismatch is visible at run start rather than discovered in logs.

File: cvs/input/config_file/inference/inferencex_atom_single/mi300x_inferencex-atom-single_deepseek-r1_fp8_smoke_config.json [hunk 2]
```diff
@@ -0,0 +1,80 @@
+{
```
Comment (atnair-amd) [inline cvs/input/config_file/inference/inferencex_atom_single/mi300x_inferencex-atom-single_deepseek-r1_fp8_smoke_config.json:1]: **[MAJOR] Config files use `ix_recipe_id` indirection instead of the flat inline structure used by vllm — and `threshold_json` is absent**

vllm configs are fully self-contained: model id, tensor parallelism, dtype, serve args, sweep, and `threshold_json` are all directly embedded. The ix-atom configs instead carry a single pointer (`ix_recipe_id`) and defer `atom_args` (e.g. `-tp 8 --kv_cache_dtype fp8 --trust-remote-code`), model id, and `bench_extra_args` to `ix_recipes.json`, resolved at load time by `apply_ix_recipe`. This means you cannot understand what a config will actually do without reading a second file.

Additionally, none of the ix-atom configs populate the `threshold_json` field (inherited from `BaseVariantConfig`). They rely on the proximity-glob fallback in `substitute_config()` which raises `ValueError` if more than one `*threshold.json` ever exists in the directory. The vllm pattern is explicit: each config carries `"threshold_json": "<path>"`.

Suggested fix: (1) flatten `atom_args` and key recipe fields inline into the config (as vllm does with `roles.server.serve_args`), and (2) add `"threshold_json": "<path>"` to each config file pointing at its paired threshold.

File: cvs/tests/inference/inferencex_atom/conftest.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+@pytest.fixture(scope="module")
+def lifecycle():
+    return InferenceLifecycle()
+
+
+@pytest.fixture(scope="module")
+def orch(cluster_dict, variant_config, lifecycle):
+    container_block = _deep_merge(
+        cluster_dict.get("container", {}),
+        orchestrator_container_from_variant(variant_config),
```
Comment (atnair-amd) [inline cvs/tests/inference/inferencex_atom/conftest.py:103]: **[NIT] `orchestrator_container_from_variant` call site has no comment explaining its side-effect**

The function now has a docstring (`includes server env`), but this call site has no annotation. A future author who copies the conftest and swaps back to `variant_config.container.model_dump()` silently drops server env vars. Suggest adding: `# also injects roles.server.env — do not replace with .container.model_dump()`

**Review by atnair-amd — COMMENTED** [review]
Automated review from five-pass analysis (structure · duplication · unit tests · code quality · live validation run).

**Blockers (must fix before merge):** false-green on missing config_file · `parse_results` error paths untested · `_client_log_failures` untested
**Majors:** driver default guard · config/threshold structure diverges from vllm · server reuse helpers untested · build_server_cmd suppression untested · `_merged_serve_args` untested · tier-explosion untested · reuse_server_across_sweep default · no ATOM early-failure detection · wheel not in shared venv
**Minors/NITs:** see inline comments

No changes are requested without reviewer approval — comments only.

File: cvs/tests/inference/inferencex_atom/inferencex_atom_single.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+def test_cell_metrics(
+    seq_combo,
+    concurrency,
+    metric_tier,
+    inf_res_dict,
+    variant_config,
+    lifecycle,
+    request,
+):
+    """One pytest row per metric tier per sweep cell (W1 gate batches)."""
```
Comment (atnair-amd) [inline cvs/tests/inference/inferencex_atom/inferencex_atom_single.py]: **[MINOR] `test_cell_metrics` failure modes are undocumented in the docstring**

There are now two `pytest.fail` paths (lines 185 and 192–195) that are stricter than vllm's silent-return behaviour: (1) no threshold specs exist for a gated tier, and (2) specs exist but none of the expected metrics are present in the actuals. Neither is documented in the docstring, making it hard for future authors to understand the contract. Suggest updating the docstring to describe both failure conditions.

File: cvs/tests/inference/inferencex_atom/conftest.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+def hf_token(variant_config):
+    path = variant_config.paths.hf_token_file
+    if not os.path.isfile(path):
+        pytest.skip(f"hf_token file missing: {path}")
+    with open(path) as fp:
+        return fp.read().strip()
+
+
+@pytest.fixture(scope="module")
+def server_session():
```
Comment (atnair-amd) [inline cvs/tests/inference/inferencex_atom/conftest.py:127]: **[MINOR] `server_session` fixture has no docstring**

This is the only fixture in `conftest.py` with no vllm analogue and its purpose is non-obvious. Suggest adding: `"""Tracks the active server session key to allow reuse across sweep cells when reuse_server_across_sweep=true."""`

File: cvs/tests/inference/inferencex_atom/inferencex_atom_single.py [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+    lifecycle,
+    request,
+):
+    if lifecycle.failed:
+        pytest.skip("a prior lifecycle stage failed")
+
+    isl = seq_combo["isl"]
+    osl = seq_combo["osl"]
+    p = variant_config.params
+    _log_variant_run_card(variant_config)
```
Comment (atnair-amd) [inline cvs/tests/inference/inferencex_atom/inferencex_atom_single.py]: **[MINOR] `_log_variant_run_card` fires once per sweep cell instead of once per variant run**

This call is inside the parametrized test body, so with N sweep cells the same run-card block is logged N times. Suggest moving it to a module-scoped autouse fixture in `conftest.py` so it fires exactly once per variant run.

File: cvs/lib/inference/inferencex_atom_orch.py [hunk 7]
```diff
@@ -0,0 +1,517 @@
...
+                    log.info("client complete (iter=%d)", it)
+                    return
+            time.sleep(self._client_poll_wait)
+        raise RuntimeError("client did not complete before poll cap")
+
+    def parse_results(self):
+        out = self.orch.exec(f"cat {shlex.quote(self._result_artifact)}")
+        results = {}
+        for host, text in out.items():
+            text = (text or "").strip()
```
Comment (atnair-amd) [inline cvs/lib/inference/inferencex_atom_orch.py:512]: **[BLOCKER] `parse_results` error paths have zero test coverage**

The empty-artifact `RuntimeError` (line 504–506) and `JSONDecodeError` branch (lines 507–512) are the most common real-world failure modes and neither has a test. `FakeOrch` (the shared canonical test double at `cvs/lib/inference/unittests/fake_orch.py`, already correctly imported by the ix-atom tests) makes both trivially exercisable:
- `FakeOrch(exec_return={"node0": ""})` → empty-artifact path
- `FakeOrch(exec_return={"node0": "not-json"})` → JSONDecodeError path

Please add two tests to `test_inferencex_atom_orch_parse.py`.

File: cvs/lib/inference/inferencex_atom_orch.py [hunk 8]
```diff
@@ -0,0 +1,517 @@
...
+        args = self._atom_client_argv() if self.driver == "atom" else self._vllm_client_argv()
+        bench_cmd = " ".join(shlex.quote(str(a)) for a in args)
+        client_cmd = f"source /tmp/server_env_script.sh && {bench_cmd} > {shlex.quote(self.client_log)} 2>&1 &"
+        self.orch.exec("bash -c " + shlex.quote(client_cmd))
+
+    def _atom_result_ready(self):
+        out = self.orch.exec(f"test -s {shlex.quote(self._result_artifact)} && echo OK || echo NO")
+        return bool(out) and all("OK" in (v or "") for v in out.values())
+
+    def _client_log_failures(self, tail_lines=2000):
```
Comment (atnair-amd) [inline cvs/lib/inference/inferencex_atom_orch.py:445]: **[BLOCKER] `_client_log_failures` is completely untested**

Three distinct behavioral outcomes have zero coverage:
1. Traceback / launch-fail crash → `CLIENT_CRASH_RE` / `CLIENT_LAUNCH_FAIL_RE` (lines 442–444) → `pytest.fail`
2. Failed-request count exceeds `bench_max_failed_requests` cap (lines 449–451) → `pytest.fail`
3. Failed-request count within cap (lines 452–457) → `log.warning`, test passes

The sentinel regexes (`FAILED_REQUESTS_RE` line 33, `CLIENT_CRASH_RE` line 34, `CLIENT_LAUNCH_FAIL_RE` lines 35–38) are also untested — a miscompiled pattern silently swallows production failures. Please add tests using `FakeOrch` with crafted log text for each outcome and test each regex explicitly.

File: cvs/tests/inference/inferencex_atom/inferencex_atom_single.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+    if rc.atom_image_pin:
+        parts.append(f"image_pin={rc.atom_image_pin}")
+    if rc.upstream_run_url:
+        parts.append(f"upstream_run={rc.upstream_run_url}")
+    if rc.notes:
+        parts.append(f"notes={rc.notes}")
+    log.info("InferenceX ATOM run card: %s", "; ".join(parts))
+
+
+def _reuse_server_flag(params) -> bool:
```
Comment (atnair-amd) [inline cvs/tests/inference/inferencex_atom/inferencex_atom_single.py]: **[MAJOR] `_reuse_server_flag` and `_server_session_key` have no unit tests**

`_reuse_server_flag` (line 53) has six meaningful inputs plus a missing-attribute edge case. `_server_session_key` (line 58): a key collision causes a server to be silently reused for a different model or shape — silent data contamination. Both are pure functions and directly importable. Please add `test_server_reuse_helpers.py` covering all `_reuse_server_flag` inputs and a `_server_session_key` collision scenario.

File: cvs/lib/inference/inferencex_atom_orch.py [hunk 10]
```diff
@@ -0,0 +1,517 @@
...
+        if self.driver == "vllm":
+            env_lines.extend(
+                [
+                    "export VLLM_USE_AITER_UNIFIED_ATTENTION=1",
+                    "export VLLM_ROCM_USE_AITER_MHA=0",
+                    "export VLLM_ROCM_USE_AITER_FUSED_MOE_A16W4=1",
+                ]
+            )
+        for k, v in self.server_env.items():
+            if k in ("CVS_GPU_MEMORY_UTIL", "VLLM_GPU_MEMORY_UTIL", "VLLM_ENFORCE_EAGER"):
```
Comment (atnair-amd) [inline cvs/lib/inference/inferencex_atom_orch.py:206]: **[MAJOR] `build_server_cmd` env-var suppression is untested**

The suppression of `CVS_GPU_MEMORY_UTIL`, `VLLM_GPU_MEMORY_UTIL`, `VLLM_ENFORCE_EAGER` (lines 205–208) is invisible under config changes. Please add a test that passes a variant with a suppressed key plus a custom key, calls `build_server_cmd()`, and asserts the suppressed key is absent while the custom key is present.

File: cvs/lib/inference/inferencex_atom_orch.py [hunk 11]
```diff
@@ -0,0 +1,517 @@
...
+        )
+        kw.update(overrides)
+        return cls(**kw)
+
+    def prepare_cell_out_dir(self):
+        """Create per-cell output directory without touching server env or cache."""
+        self.orch.exec(f"mkdir -p {shlex.quote(self.out_dir)}")
+
+    @classmethod
+    def _merged_serve_args(cls, variant):
```
Comment (atnair-amd) [inline cvs/lib/inference/inferencex_atom_orch.py:167]: **[MAJOR] `_merged_serve_args` gpu-memory-util promotion is untested**

The promotion of `CVS_GPU_MEMORY_UTIL` from `server_env` into `gpu-memory-utilization` in the serve-args dict (lines 172–174) is a separate code path from `orchestrator_container_from_variant` and has no test. Please add a test asserting the promotion happens when the key is absent from `merged`, and is a no-op when already present.

File: cvs/tests/inference/inferencex_atom/inferencex_atom_single.py [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+
+
+def pytest_generate_tests(metafunc):
+    config_file = metafunc.config.getoption("config_file")
+    if not config_file or not os.path.isfile(config_file):
+        return
+    with open(config_file) as fp:
+        raw = json.load(fp)
+    cases, ids = expand_sweep(raw.get("sweep", {}))
+    if "metric_tier" in metafunc.fixturenames:
```
Comment (atnair-amd) [inline cvs/tests/inference/inferencex_atom/inferencex_atom_single.py]: **[MAJOR] `pytest_generate_tests` tier-explosion path is untested**

The `"metric_tier" in metafunc.fixturenames` branch (line 84) and the tier × sweep multiplication (lines 85–92) are a separate code path from `expand_sweep` (which is tested). The `len(cases) * len(METRIC_TIER_ORDER)` product is not pinned anywhere. Please add a test using a `SimpleNamespace` fake metafunc (with a recording `.parametrize`) asserting the correct number of parametrize calls and IDs for a known sweep config.

File: cvs/tests/inference/inferencex_atom/inferencex_atom_single.py [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+
+def _tier_display_metric(tier):
+    if tier == "record":
+        return RECORD_METRICS[0] if RECORD_METRICS else "output_throughput"
+    names = METRIC_TIERS.get(tier, ())
+    return names[0] if names else tier
+
+
+def pytest_generate_tests(metafunc):
+    config_file = metafunc.config.getoption("config_file")
```
Comment (atnair-amd) [inline cvs/tests/inference/inferencex_atom/inferencex_atom_single.py:46]: **[BLOCKER] Missing `config_file` produces a false-green zero-test run**

Lines 78–79: `if not config_file or not os.path.isfile(config_file): return` causes pytest to collect 0 tests and exit 0 — indistinguishable from a fully passing run. This is a silent failure mode: a typo in `--config_file` or a missing file will look like success in CI. Please replace the `return` with `pytest.exit("--config_file not found or not specified", returncode=1)` or raise `pytest.UsageError`.

File: cvs/tests/inference/inferencex_atom/inferencex_atom_single.py [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+        parts.append(f"image_pin={rc.atom_image_pin}")
+    if rc.upstream_run_url:
+        parts.append(f"upstream_run={rc.upstream_run_url}")
+    if rc.notes:
+        parts.append(f"notes={rc.notes}")
+    log.info("InferenceX ATOM run card: %s", "; ".join(parts))
+
+
+def _reuse_server_flag(params) -> bool:
+    raw = str(getattr(params, "reuse_server_across_sweep", "true")).strip().lower()
```
Comment (atnair-amd) [inline cvs/tests/inference/inferencex_atom/inferencex_atom_single.py]: **[MAJOR] `_reuse_server_flag` defaults to `"true"` when `reuse_server_across_sweep` is absent from config**

`getattr(params, "reuse_server_across_sweep", "true")` — a config that omits the field silently enables server reuse, which is a surprising default (especially for smoke runs). The same default is set at `inferencex_atom_config_loader.py` line 68. Suggest changing to `"false"` (safe: always restart) or making the field required in the schema.

File: cvs/lib/inference/inferencex_atom_orch.py [hunk 15]
```diff
@@ -0,0 +1,517 @@
...
+            f"grep -qiE {shlex.quote(pattern)} {shlex.quote(self.server_log)}",
+            detailed=True,
+        )
+        return bool(out) and all(r["exit_code"] == 0 for r in out.values())
+
+    def wait_ready(self):
+        log.info("waiting %ds for server log to materialise", self._precheck_wait)
+        time.sleep(self._precheck_wait)
+
+        if self.driver != "atom":
```
Comment (atnair-amd) [inline cvs/lib/inference/inferencex_atom_orch.py]: **[MAJOR] No early-failure detection for the ATOM driver during the `wait_ready` poll loop**

Lines 294–298: the `EARLY_FAILURE_RE` log scan is gated on `if self.driver != "atom"` — the atom driver skips it entirely. During the smoke run the server was silent for 29 minutes with 28 consecutive health-poll `NO` responses; a crashed server would be indistinguishable from a loading one until timeout. `start_server` does check `EARLY_FAILURE_RE` on the immediate nohup shell output (line 257) but that only catches shell-level launch failures, not server-side startup failures written to the log file during the load window. Suggest adding `EARLY_FAILURE_RE` pattern matching against the ATOM server log inside the `wait_ready` poll loop, analogous to the vllm path.

File: cvs/lib/inference/utils/vllm_benchmark_scripts/__init__.py [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+        '[ -n "$BENCH_PY" ] && [ -n "$BENCH_KIND" ] && break; '
+        "done; "
+        '[ -n "$BENCH_KIND" ] && [ -n "$BENCH_PY" ] || '
+        "{ echo 'CVS: could not resolve vLLM benchmark driver (tried python3.13..python3)' >&2; exit 1; }; "
+        'if [ "$BENCH_KIND" = script ] && { [ ! -f "$BENCH_SCRIPT" ] || [ ! -r "$BENCH_SCRIPT" ]; }; then '
+        '{ echo "CVS: BENCH_SCRIPT is missing or unreadable: $BENCH_SCRIPT" >&2; exit 1; }; fi; '
+        '_cvs_run_bench() { case "$BENCH_KIND" in cli_module) '
+        '"$BENCH_PY" -m vllm.entrypoints.cli.main bench serve "$@" '
+        ';; *) "$BENCH_PY" "$BENCH_SCRIPT" "$@" ;; esac; }; '
+        "export BENCH_PY BENCH_KIND BENCH_SCRIPT"
```
Comment (amd-droy) [inline cvs/lib/inference/utils/vllm_benchmark_scripts/__init__.py:199]: This function returns a long shell string containing embedded Python and may create several problems.
# RECOMMENDED 
PYTHON_CANDIDATES = ["python3.13", "python3.12", "python3.11", "python3.10", "python3"]

def _build_interpreter_loop(py_q: str, candidates: list[str]) -> str:
    """Build the shell interpreter-discovery loop separately for testability."""
    candidate_str = " ".join(candidates)
    return (
        f"for _cvs_py in {candidate_str}; do\n"
        f"  _cvs_eval=$($_cvs_py -c {py_q}) || continue\n"
        ...
    )

File: cvs/lib/inference/utils/inferencex_atom_parsing.py [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+CLIENT_METRICS: list[tuple[str, str]] = []
+_inserted = False
+for _name, _unit in _VLLM_CLIENT_METRICS:
+    CLIENT_METRICS.append((_name, _unit))
+    if _name == "per_gpu_throughput" and not _inserted:
+        CLIENT_METRICS.append(_METRIC_INSERT)
+        _inserted = True
+if not _inserted:
+    CLIENT_METRICS.append(_METRIC_INSERT)
+CLIENT_METRIC_UNITS = dict(CLIENT_METRICS)
```
Comment (amd-droy) [inline cvs/lib/inference/utils/inferencex_atom_parsing.py:32]: # Alternative: 
idx = next((i for i, (n, _) in enumerate(_VLLM_CLIENT_METRICS) if n == "per_gpu_throughput"), None)
CLIENT_METRICS = list(_VLLM_CLIENT_METRICS)
CLIENT_METRICS.insert((idx + 1) if idx is not None else len(CLIENT_METRICS), _METRIC_INSERT)

File: cvs/lib/inference/inference_suite_lifecycle.py [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+    saw_marker = False
+    for text in (out or {}).values():
+        text = (text or "").strip()
+        if text == "__DU_ERROR__":
+            return None
+        if text == "__MISSING__":
+            saw_marker = True
+            continue
+        if text.isdigit():
+            total = max(total, int(text))
```
Comment (amd-droy) [inline cvs/lib/inference/inference_suite_lifecycle.py]: Taking max across hosts means the return value is the largest single-host measurement, not the total bytes across the distributed model cache.  I think it should be 
  if text.isdigit():
        total += int(text)   # sum shards, not max shard

File: cvs/lib/inference/inference_suite_lifecycle.py [hunk 19]
```diff
@@ -0,0 +1,245 @@
...
+
+def attach_lifecycle_html_table(item, report):
+    if report.when != "call":
+        return
+    lc = item.funcargs.get("lifecycle")
+    rows = getattr(lc, "report", {}).get(item.nodeid) if lc else None
+    if not rows:
+        return
+    try:
+        import pytest_html
```
Comment (amd-droy) [inline cvs/lib/inference/inference_suite_lifecycle.py]: you can move this to the top in the import section

File: cvs/lib/inference/utils/vllm_benchmark_scripts/vllm_serve_mi300x.sh [hunk 20]
```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/env bash
```
Comment (atnair-amd) [inline cvs/lib/inference/utils/vllm_benchmark_scripts/vllm_serve_mi300x.sh:1]: noticed this: file path should be cvs/lib/inference/utils/vllm_benchmark_scripts/*

**Review by amd-droy — APPROVED** [review]
looks good to me. thanks @hnimra-amd

**Review by atnair-amd — APPROVED** [review]
lgtm @hnimra-amd
